nx/lib/nx.ex

```diff
@@ -6612,6 +6612,152 @@ defmodule Nx do
   defp mean_den(shape, [axis | axes]) when axis >= 0,
     do: elem(shape, axis) * mean_den(shape, axes)
 
+  @doc """
+  Returns the weighted mean for the tensor and the weights.
+
+  If the `:axis` option is given, it aggregates over
+  that dimension, effectively removing it. `axes: [0]`
+  implies aggregating over the highest order dimension
+  and so forth. If the axis is negative, then counts
+  the axis from the back. For example, `axis: [-1]` will
+  always aggregate all rows. You can only aggregate over
+  one axis.
+
+  You may optionally set `:keep_axis` to true, which will
+  retain the rank of the input tensor by setting the averaged
+  axis to size 1.
+
+  ## Examples
+
+      iex> Nx.weighted_mean(Nx.tensor(42), Nx.tensor(2))
+      #Nx.Tensor<
+        f32
+        42.0
+      >
+
+      iex> Nx.weighted_mean(Nx.tensor([1, 2, 3]), Nx.tensor([3, 2, 1]))
+      #Nx.Tensor<
+        f32
+        1.6666666269302368
+      >
+
+  ### Aggregating over an axis
+
+      iex> Nx.weighted_mean(Nx.tensor([1, 2, 3], names: [:x]), Nx.tensor([4, 5, 6]), axis: 0)
+      #Nx.Tensor<
+        f32
+        2.133333444595337
+      >
+
+      iex> Nx.weighted_mean(Nx.tensor([1,2,3], type: :u8, names: [:x]), Nx.tensor([1,3,5]), axis: :x)
+      #Nx.Tensor<
+        f32
+        2.444444417953491
+      >
+
+      iex> t = Nx.tensor([[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]]], names: [:x, :y, :z])
+      iex> weights = Nx.tensor([[[8, 2, 4], [0, 3, 5]], [[4, 2, 6], [3, 1, 5]]])
+      iex> Nx.weighted_mean(t, weights, axis: :x)
+      #Nx.Tensor<
+        f32[y: 2][z: 3]
+        [
+          [3.0, 5.0, 6.599999904632568],
+          [10.0, 6.5, 9.0]
+        ]
+      >
+
+      iex> t = Nx.tensor([[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]]], names: [:x, :y, :z])
+      iex> weights = Nx.tensor([[[8, 2, 4], [0, 3, 5]], [[4, 2, 6], [3, 1, 5]]])
+      iex> Nx.weighted_mean(t, weights, axis: -1)
+      #Nx.Tensor<
+        f32[x: 2][y: 2]
+        [
+          [1.7142857313156128, 5.625],
```

Comment:
I think we should have cleaner numbers in the example.
I suggest having a weights tensor with lots of 1s, -1s and 0s so that it's easier to calculate everything